Added support for executing remote commands through the command line.

dov opened the pull request:
* Added text flag to ws.write for sending text (terminal) messages.
* Rewrite of help.
* Rewrite of parsing.

pfalcon commented on webrepl_cli.py:
```diff
@@ -36,13 +36,14 @@ def __init__(self, s):
 
-        def write(self, data):
+        def write(self, data, text=False):
+            OpCode = 1 if text else 2
```
Codestyle: you violate both PEP8 and convention of this file (which of course uses standard_python_style).

pfalcon commented on webrepl_cli.py:
```diff
@@ -36,13 +36,14 @@ def __init__(self, s):
             if l < 126:
                 # TODO: hardcoded "binary" type
-                hdr = struct.pack(">BB", 0x82, l)
+                hdr = struct.pack(">BB", 0x80+OpCode, l)
```
PEP8 violation: spacing around operators.

pfalcon commented on webrepl_cli.py:
```diff
@@ -164,15 +165,32 @@ def get_file(ws, local_file, remote_file):
-    print("  %s script.py 192.168.4.1:/another_name.py" % exename)
-    print("  %s script.py 192.168.4.1:/app/" % exename)
-    print("  %s 192.168.4.1:/app/script.py ." % exename)
+    exename = os.path.basename(sys.argv[0])
```
> Rewrite of help.

That maybe even nice (maybe not), but definitely can't be part of a commit with the description "Added support for executing remote commands through the command line.". One commit - one logical change. Each change should be grounded (i.e., there should be arguments why change is needed).

pfalcon commented:
Thanks for the submission, but the patch has multiple issue. Are you aware that there're other tools which already implement (should implement IIRC) this functionality?

dov commented:
Thanks for the comments, and for webrepl! I admit that the changes in this pull request was more of a sloppy hack for learning how to communicate with the esp8266, but I thought that there is no harm in trying to contribute it back. Let me know if you think whether you think my patch in within the scope of how you envision webrepl_cli and I'll make the effort to clean it up. If not, that's fine with me as well.

pfalcon commented:
We, MicroPython maintainers, are trying to concentrate on the core interpreter and libraries. So, as long as there're community utility/tooling projects, our preference is to save maintenance efforts on our side, and vice-versa, encourage their further development and maintenance on community side. So, I'd really recommend to see if there's a tool which already suits your needs among those listed in https://forum.micropython.org/viewtopic.php?f=16&t=1908 (and on that forum in general), and if you see opportunity to improve them/it, to contribute to them instead.

dov commented:
I understand and I accept this policy, but I have a few comments:

* If the purpose of webrepl_cli.py is to copy, then its name is misleading, and it should more aptly be called e.g. webrepl_copy.py .
* Providing the password in the command line is in my opinion an appropriate feature within the limited scope that you defined.
*  Since you protect the copy functionality from the communications functionality through `if __name__=='__main__'`, it invites for using the library as an import for other (non-core) python code. In such a case the ascii option that I added is missing in order to send webrepl commands.

If I were to address the latter two issues in code style complient separate pull requests, would that be more acceptable?

roger- commented:
Is there another terminal client with this feature? I’d love to see a version of this merged.

dov commented:
I just updated my master branch, and added the following functionality:

- Placed all the connection functionality into a class for easy reuse.
- Made the command option return the micropython webrepl output. This means that it can be used as a simple RPC protocol. E.g. the following command line returns the input value of D8:

```
$ ./webrepl_cli.py --passwd secret --host 192.168.1.220 --cmd 'import machine; machine.Pin(15,machine.Pin.IN).value()'
0
```

Since my previous pull request was ignored without explanation during the last four years, I doubt that there will be any interest in merging this either. If there were, I'd separate this change into at least three separate commits:

1. Rewrite of the help
2. Introduction of the `WebreplCLI` class
3. Addition of return value for the `command()` method.

dov commented:
I closed this pull request, in order to free up my master branch for additional work. If there is any interest, I'll be happy to open a new pull request based on a different branch.
